File: interpreter.cpp
#include "interpreter.hpp"
#include "environment.hpp"
#include "expr.hpp"
#include "leaf_error.hpp"
#include "leaf_object.hpp"
#include "leaf_function.hpp"
#include "leaf_struct.hpp"
#include "stmt.hpp"
#include "token.hpp"
#include "token_type.hpp"
#include "tools/binary_operations.hpp"

#include <cstdlib>
#include <format>

using namespace std::string_literals;

using enum TokenType;


// Break
class Break { };

// Continue
class Continue { };


// Interpreter
Interpreter::Interpreter(Environment* environment) :
    m_environment { environment }
    {
    }

Interpreter::~Interpreter() {
    Environment::delete_object(m_environment);
}


// Private methods
auto Interpreter::execute_stmt(Stmt* stmt) -> void {
    stmt->accept(this);
}

auto Interpreter::evaluate(Expr* expr) -> LeafObject* {
    return expr->accept(this);
}


// Public methods
auto Interpreter::execute(std::vector<Stmt*>& statements) -> void {
    for (Stmt* stmt : statements) {
        execute_stmt(stmt);
    }
}

auto Interpreter::execute(std::vector<Expr*>& expressions) -> LeafObject* {
    for (Expr* expr : expressions) {
        if (expr->type() != ExprType::k_null) {
            std::cout << evaluate(expr) << "\n";
        }
    }
    return nullptr;
}


auto Interpreter::visit_printstmt(PrintStmt* stmt) -> void {
    LeafObject* value { evaluate(stmt->expr) };
    std::cout << value;
}

auto Interpreter::visit_printlnstmt(PrintlnStmt* stmt) -> void {
    LeafObject* value { evaluate(stmt->expr) };
    std::cout << value << "\n";
}

auto Interpreter::visit_expressionstmt(ExpressionStmt* stmt) -> void {
    evaluate(stmt->expr);
}

auto Interpreter::visit_varstmt(VarStmt* stmt) -> void {
    m_environment->insert_var(stmt->identifier, evaluate(stmt->expr));
}

auto Interpreter::visit_conststmt(ConstStmt* stmt) -> void {
    m_environment->insert_const(stmt->identifier, evaluate(stmt->expr));
}

auto Interpreter::visit_blockstmt(BlockStmt* stmt, Environment* environment) -> void {
    environment->set_parent(m_environment);
    m_environment = environment;

    try {
        for (Stmt* statement : stmt->statements) {
            execute_stmt(statement);
        }
    } catch (LeafObject* return_value) {
        m_environment = environment->parent();
        Environment::delete_object(environment);
        throw return_value;
    }
    m_environment = environment->parent();
    Environment::delete_object(environment);
}

auto Interpreter::visit_ifstmt(IfStmt* stmt) -> void {
    for (auto statement : stmt->statements) {
        if (statement.first == nullptr) {
            execute_stmt(statement.second);
            break;
        }
        LeafObject* condition { evaluate(statement.first) };
        if (condition->is_truthy()) {
            execute_stmt(statement.second);
            break;
        }
    }
}

auto Interpreter::visit_forstmt(ForStmt* stmt) -> void {
    BlockCtx prev_block_ctx { m_block_ctx };
    m_block_ctx = BlockCtx::k_loop;
    try {
        if (stmt->condition == nullptr) {
            while (true) {
                try {
                    execute_stmt(stmt->block_stmt);
                } catch (Continue&) {
                }
            }
            return;
        }

        while (evaluate(stmt->condition)->is_truthy()) {
            try {
                execute_stmt(stmt->block_stmt);
            } catch (Continue&) {
            }
        }
    } catch (Break&) {
    }
    m_block_ctx = prev_block_ctx;
}

auto Interpreter::visit_breakstmt(BreakStmt* stmt) -> void {
    if (m_block_ctx == BlockCtx::k_loop) {
        throw Break { };
        return;
    }
    LeafError::instance()->runtime_error(
        stmt->line,
        "'break' can only be used within the context of a loop."s
    );
}

auto Interpreter::visit_continuestmt(ContinueStmt* stmt) -> void {
    if (m_block_ctx == BlockCtx::k_loop) {
        if (stmt->step_expr) {
            evaluate(stmt->step_expr);
        }
        throw Continue { };
        return;
    }
    LeafError::instance()->runtime_error(
        stmt->line,
        "'continue' can only be used within the context of a loop."s
    );
}

auto Interpreter::visit_returnstmt(ReturnStmt* stmt) -> void {
    if (m_called_ctx == CalledCtx::k_function) {
        throw evaluate(stmt->value);
    }
    if (m_called_ctx == CalledCtx::k_constructor) {
        LeafError::instance()->runtime_error(
            stmt->token->line(), "'return' statement can not be used inside constructor."s);
    }
    LeafError::instance()->runtime_error(
        stmt->token->line(), "'return' statement can only be used within the context of a function."s);
}


auto Interpreter::visit_nullexpr(NullExpr*) -> LeafObject* {
    return LeafNull::create_object();
}

auto Interpreter::visit_assignexpr(AssignExpr* expr) -> LeafObject* {
    Token* left { expr->left };
    if (left->type() != k_identifier) {
        LeafError::instance()->runtime_error(expr->oper->line(), expr->oper->lexeme(), "Left hand operator of '=' must be a variable."s);
    }
    LeafObject* right { evaluate(expr->right) };

    if (m_environment->get_qualifier(left) == "const"s) {
        LeafError::instance()->runtime_error(
            expr->oper->line(),
            expr->oper->lexeme(),
            std::format("Const variable '{}' can not be re-assigned a value.", left->lexeme())
        );
    }

    m_environment->assign(left, right);
    return const_cast<LeafObject*>(right);
}

auto Interpreter::visit_ternaryexpr(TernaryExpr* expr) -> LeafObject* {
    LeafObject* condition { evaluate(expr->condition) };
    if (condition->is_truthy()) {
        return evaluate(expr->first);
    }
    return evaluate(expr->second);
}

auto Interpreter::visit_binaryexpr(BinaryExpr* expr) -> LeafObject* {
    auto left { evaluate(expr->left) };
    auto right { evaluate(expr->right) };
    return BinaryOperations::instance()->execute(expr->oper, left, right);
}

auto Interpreter::visit_unaryexpr(UnaryExpr* expr) -> LeafObject* {
    auto operand { evaluate(expr->operand) };
    if (expr->oper->lexeme() == "-") {
        if (operand->type() != ObjectType::k_number) {
            LeafError::instance()->runtime_error(expr->oper->line(), expr->oper->lexeme(), "Operand must be a number."s);
        }
        return LeafNumber::create_object(-dynamic_cast<LeafNumber*>(operand)->value());
    } else {
        return LeafBool::create_object( !operand->is_truthy() );
    }
}

auto Interpreter::visit_exponentexpr(ExponentExpr* expr) -> LeafObject* {
    auto left { evaluate(expr->left) };
    auto right { evaluate(expr->right) };
    return BinaryOperations::instance()->execute(expr->oper, left, right);
}

auto Interpreter::visit_groupingexpr(GroupingExpr* expr) -> LeafObject* {
    return evaluate(expr->expression);
}

auto Interpreter::visit_primaryexpr(PrimaryExpr* expr) -> LeafObject* {
    if (expr == nullptr) {
        return nullptr;
    }

    auto type { expr->token->type() };

    if (type == k_number) {
        return LeafNumber::create_object(expr->token);
    } else if (type == k_string) {
        return LeafString::create_object(expr->token);
    } else if (type == k_true || type == k_false) {
        return LeafBool::create_object(expr->token);
    } else if (type == k_null) {
        return LeafNull::create_object(expr->token);
    } else if (type == k_identifier) {
        return const_cast<LeafObject*>(m_environment->get(expr->token));
    }
    return nullptr;
}

auto Interpreter::visit_functionexpr(FunctionExpr* expr) -> LeafObject* {
    return LeafFunction::create_object(expr->parameters, expr->block_stmt, m_environment->make_closure());
}

auto Interpreter::visit_callexpr(CallExpr* expr) -> LeafObject* {
    LeafObject* value { };

    if (expr->identifier) {
        if (m_is_method) {
            value = m_current_instance->value().at(expr->identifier->lexeme());
        } else {
            value = m_environment->get(expr->identifier);
        }
        if (value->type() != ObjectType::k_function && value->type() != ObjectType::k_struct) {
            LeafError::instance()->runtime_error(
                expr->identifier->line(),
                std::format("'{}' is not callable.", expr->identifier->lexeme()));
        }
    } else if (expr->expr) {
        value = evaluate(expr->expr);
    }

    CalledCtx prev_ctx { m_called_ctx };
    LeafStructInstance* previous_instance { m_current_instance };

    LeafFunction* casted_value { };
    if (value->type() == ObjectType::k_struct) {
        m_called_ctx = CalledCtx::k_constructor;

        casted_value = m_environment->metaclass_get_method(expr->identifier, "__construct"s);
        if (casted_value == nullptr) {
            LeafError::instance()->runtime_error(
                expr->identifier->line(),
                std::format("No constructor is provided for struct '{}'", expr->identifier->lexeme())
            );
        }

        LeafStruct* leaf_struct { dynamic_cast<LeafStruct*>(value) };

        std::unordered_map<std::string, LeafObject*> fields { };
        for (Token* field : leaf_struct->value()) {
            fields.insert_or_assign(field->lexeme(), LeafNull::create_object());
        }
        m_current_instance = LeafStructInstance::create_object(leaf_struct->identifier, fields);

    } else if (value->type() == ObjectType::k_function) {
        m_called_ctx = CalledCtx::k_function;
        casted_value = dynamic_cast<LeafFunction*>(value);
    }

    if (casted_value->parameters.size() != expr->arguments.size()) {
        LeafError::instance()->runtime_error(
            expr->identifier->line(),
            std::format("Number of arguments does not match number of parameters for function '{}'.", expr->identifier->lexeme()));
    }

    LeafObject* return_value { casted_value->call(expr->arguments, this) };

    m_called_ctx = prev_ctx;
    m_current_instance = previous_instance;
    return return_value;
}

auto Interpreter::visit_structexpr(StructExpr* expr) -> LeafObject* {
    LeafStruct* leaf_struct { LeafStruct::create_object(expr->identifier, expr->fields) };
    m_environment->insert_const(expr->identifier, leaf_struct);

    for (FunctionExpr* func_expr : expr->methods) {
        m_environment->metaclass_add_method(expr->identifier, func_expr->identifier->lexeme(), dynamic_cast<LeafFunction*>(evaluate(func_expr)));
    }
    return leaf_struct;
}

auto Interpreter::visit_getexpr(GetExpr* expr) -> LeafObject* {
    if (expr->left) {
        LeafObject* value { m_environment->get(expr->left) };
        return value;
    }

    LeafObject* left { evaluate(expr->left_expr) };

    std::string member { };
    if (expr->right_expr == nullptr) {
        member.assign("Field");
    } else {
        member.assign("Method");
    }
    if (left->type() != ObjectType::k_struct_instance) {
        LeafError::instance()->runtime_error(
            expr->right->line(),
            std::format("{} '{}' can not be called on a non-struct value.", member, expr->right->lexeme())
        );
    }
    LeafStructInstance* instance { dynamic_cast<LeafStructInstance*>(left) };
    if (instance->value().contains(expr->right->lexeme()) == false) {
        LeafError::instance()->runtime_error(
            expr->right->line(),
            std::format("{} '{}' does not exist on given struct value.", member, expr->right->lexeme()));
    }

    if (expr->right_expr == nullptr) {
        return instance->value().at(expr->right->lexeme());
    }

    auto is_method_snap { m_is_method };
    m_is_method = true;

    auto current_instance_snap { m_current_instance };
    m_current_instance = instance;

    LeafObject* result { evaluate(expr->right_expr) };

    m_current_instance = current_instance_snap;
    m_is_method = is_method_snap;
    return result;
}

auto Interpreter::visit_setexpr([[maybe_unused]] SetExpr* expr) -> LeafObject* {
    LeafObject* value { evaluate(expr->left->left_expr) };
    if (value->type() != ObjectType::k_struct_instance) {
        LeafError::instance()->runtime_error(
            expr->left->right->line(),
            std::format("Field '{}' can not be called on non-struct value.", expr->left->right->lexeme()));
    }
    LeafStructInstance* instance { dynamic_cast<LeafStructInstance*>(value) };
    if (instance->value().contains(expr->left->right->lexeme()) == false) {
        LeafError::instance()->runtime_error(
            expr->left->right->line(),
            std::format("Field '{}' does not exist on given struct value.", expr->left->right->lexeme()));
    }
    LeafObject* value_to_assign { evaluate(expr->right) };
    instance->value().insert_or_assign(
        expr->left->right->lexeme(), value_to_assign);
    return value_to_assign;
}

File: lib/interpreter.ts
import binaryOperation from "./binaryOperation"
import Callable, { LeafFunction, LeafMethod, ReturnValue } from "./callable"
import LeafClass, { LeafInstance } from "./class"
import Environment from "./environment"
import LeafError from "./error"
import IExpr, {
    AssignExpr,
    BinaryExpr,
    CallExpr,
    ClassExpr,
    ExponentExpr,
    FuncExpr,
    GetExpr,
    GroupingExpr,
    IdentifierExpr,
    IExprVisitor,
    LiteralExpr,
    NilExpr,
    SetExpr,
    TernaryExpr,
    ThisExpr,
    UnaryExpr,
} from "./expr"
import IObj, { ObjBool, ObjNil, ObjNumber, ObjString } from "./object"
import IStmt, {
    BlockStmt,
    BreakStmt,
    ClassStmt,
    ConstStmt,
    ContinueStmt,
    ExprStmt,
    ForStmt,
    ForWrapperStmt,
    FuncStmt,
    IfStmt,
    InfiniteForStmt,
    IStmtVisitor,
    LetForStmt,
    LetStmt,
    PrintlnStmt,
    PrintStmt,
    ReturnStmt,
    WhileForStmt,
} from "./stmt"
import Token from "./token"
import tokenType from "./tokenType"
import utils from "./utils"

enum BlockCtx {
    NONE,
    FOR,
}

enum CallableCtx {
    NONE,
    FUNC,
}

const breakStmt = "breakStmt"
const continueStmt = "continueStmt"

export default class Interpreter implements IExprVisitor<IObj>, IStmtVisitor {
    private constructor(
        private _environment = Environment.createInstance(),
        private _blockCtx = BlockCtx.NONE,
        private _callableCtx = CallableCtx.NONE
    ) {}

    static createInstance() {
        return new Interpreter()
    }

    environment() {
        return this._environment
    }

    setEnvironment(environment: Environment) {
        this._environment = environment
    }

    instance() {
        return this._environment.instance()
    }

    setInstance(instance: LeafInstance) {
        this._environment.setInstance(instance)
    }

    execute(stmts: IStmt[]): void {
        stmts.forEach((stmt) => {
            this.executeStmt(stmt)
        })
    }

    // statement visitor functions
    private executeStmt(stmt: IStmt): void {
        stmt.accept(this)
    }

    visitPrintStmt(stmt: PrintStmt): void {
        process.stdout.write(`${this.evaluate(stmt.expr()).value()}`)
    }

    visitPrintlnStmt(stmt: PrintlnStmt): void {
        console.log(this.evaluate(stmt.expr()).value())
    }

    visitLetStmt(stmt: LetStmt): void {
        this._environment.insertLet(
            stmt.identifier(),
            this.evaluate(stmt.expr())
        )
    }

    visitConstStmt(stmt: ConstStmt): void {
        this._environment.insertConst(
            stmt.identifier(),
            this.evaluate(stmt.expr())
        )
    }

    visitBlockStmt(stmt: BlockStmt): void {
        const environment = Environment.createInstance(this._environment)
        this._environment = environment

        try {
            stmt.stmts().forEach((stmt) => {
                this.executeStmt(stmt)
            })
        } catch (err) {
            this._environment = this._environment.parent() as Environment
            throw err
        }
        this._environment = this._environment.parent() as Environment
    }

    visitIfStmt(stmt: IfStmt): void {
        stmt.stmtTable().find(([condition, block]) => {
            if (utils.isTruthy(this.evaluate(condition))) {
                this.executeStmt(block)
                return true
            }
            return false
        })
    }

    visitForWrapperStmt(stmt: ForWrapperStmt): void {
        const prevBlockCtx = this._blockCtx
        this._blockCtx = BlockCtx.FOR
        try {
            this.executeStmt(stmt.forStmt())
        } catch (err) {
            if ((err as Error).message !== breakStmt) {
                throw err
            }
        }
        this._blockCtx = prevBlockCtx
    }

    visitInfiniteForStmt(stmt: InfiniteForStmt): void {
        while (true) {
            this.continueWrapper(stmt.stmts())
        }
    }

    visitWhileForStmt(stmt: WhileForStmt): void {
        while (utils.isTruthy(this.evaluate(stmt.condition()))) {
            this.continueWrapper(stmt.stmts())
        }
    }

    visitForStmt(stmt: ForStmt): void {
        for (
            this.evaluate(stmt.init());
            utils.isTruthy(this.evaluate(stmt.condition()));
            this.evaluate(stmt.step())
        ) {
            this.continueWrapper(stmt.stmts())
        }
    }

    visitLetForStmt(stmt: LetForStmt): void {
        const environment = Environment.createInstance(this._environment)
        this._environment = environment

        try {
            for (
                this.executeStmt(stmt.init());
                utils.isTruthy(this.evaluate(stmt.condition()));
                this.evaluate(stmt.step())
            ) {
                this.continueWrapper(stmt.stmts())
            }
        } catch (err) {
            this._environment = this._environment.parent() as Environment
            if ((err as Error).message !== breakStmt) {
                throw err
            }
        }
    }

    visitBreakStmt(stmt: BreakStmt): void {
        if (this._blockCtx !== BlockCtx.FOR) {
            throw LeafError.getInstance().makeRuntimeError(
                stmt.keyword().line(),
                "'break' can only be used inside a loop"
            )
        }
        throw new Error(breakStmt)
    }

    visitContinueStmt(stmt: ContinueStmt): void {
        if (this._blockCtx !== BlockCtx.FOR) {
            throw LeafError.getInstance().makeRuntimeError(
                stmt.keyword().line(),
                "'continue' can only be used inside a loop"
            )
        }
        throw new Error(continueStmt)
    }

    visitFuncStmt(stmt: FuncStmt): void {
        this._environment.insertConst(
            stmt.name() as Token,
            this.evaluate(stmt.expr())
        )
    }

    visitReturnStmt(stmt: ReturnStmt): void {
        if (this._callableCtx === CallableCtx.NONE) {
            throw LeafError.getInstance().makeRuntimeError(
                stmt.keyword().line(),
                "'return' statement can only be used inside a function"
            )
        }
        throw ReturnValue.createInstance(this.evaluate(stmt.value()))
    }

    visitClassStmt(stmt: ClassStmt): void {
        this._environment.insertConst(stmt.name(), this.evaluate(stmt.expr()))
    }

    visitExprStmt(stmt: ExprStmt): void {
        this.evaluate(stmt.expr())
    }

    // expression visitor functions
    evaluate(expr: IExpr) {
        return expr.accept(this)
    }

    visitAssignExpr(expr: AssignExpr): IObj {
        const value = this.evaluate(expr.right())
        this._environment.set(expr.left().token(), value)
        return value
    }

    visitTernaryExpr(expr: TernaryExpr): IObj {
        const condition = this.evaluate(expr.condition())
        if (utils.isTruthy(condition)) {
            return this.evaluate(expr.first())
        } else {
            return this.evaluate(expr.second())
        }
    }

    visitBinaryExpr(expr: BinaryExpr): IObj {
        const oper = expr.oper()
        const left = this.evaluate(expr.left())
        const right = this.evaluate(expr.right())
        return binaryOperation(oper, left, right)
    }

    visitExponentExpr(expr: ExponentExpr): IObj {
        const oper = expr.oper()
        const left = this.evaluate(expr.left())
        const right = this.evaluate(expr.right())
        return binaryOperation(oper, left, right)
    }

    visitUnaryExpr(expr: UnaryExpr): IObj {
        const oper = expr.oper()
        const expr_ = this.evaluate(expr.expr())
        if (oper.lexeme() === "!") {
            return ObjBool.createInstance(!utils.isTruthy(expr_))
        }
        if (expr_ instanceof ObjNumber) {
            return ObjNumber.createInstance(-expr_.value())
        }
        throw LeafError.getInstance().makeRuntimeError(
            oper.line(),
            "operand of unary '-' can only be a number;"
        )
    }

    visitFuncExpr(expr: FuncExpr): IObj {
        return LeafFunction.createInstance(
            expr.parameters(),
            expr.stmts(),
            this._environment
        )
    }

    visitCallExpr(expr: CallExpr): IObj {
        const caller = this.evaluate(expr.caller())

        if (!(caller instanceof Callable)) {
            if (caller instanceof LeafClass) {
                return this.callClass(caller, expr)
            }

            throw LeafError.getInstance().makeRuntimeError(
                this.getCallLine(expr),
                "expression is not callable"
            )
        }

        const prevCtx = this._callableCtx
        this._callableCtx = CallableCtx.FUNC

        const func = caller as LeafFunction

        const value = func.call(expr.args(), this, expr)

        this._callableCtx = prevCtx
        return value
    }

    visitGetExpr(expr: GetExpr): IObj {
        const caller = this.evaluate(expr.caller())

        if (!(caller instanceof LeafInstance)) {
            throw LeafError.getInstance().makeRuntimeError(
                expr.name().line(),
                `'${expr
                    .name()
                    .lexeme()}' can not be called on non class instance`
            )
        } else {
            if (caller.fields().has(expr.name().lexeme())) {
                return caller.fields().get(expr.name().lexeme()) as IObj
            }

            const leafClass = this._environment.get(
                caller.className()
            ) as LeafClass

            if (leafClass.methods().has(expr.name().lexeme())) {
                const method = leafClass
                    .methods()
                    .get(expr.name().lexeme()) as LeafMethod
                method.setInstance(caller)
                return method as IObj
            }
        }
        throw LeafError.getInstance().makeRuntimeError(
            this.getCallLine(expr),
            `name '${expr.name().lexeme()}' isn't defined on caller`
        )
    }

    visitSetExpr(expr: SetExpr): IObj {
        const caller = this.evaluate(expr.caller())

        if (!(caller instanceof LeafInstance)) {
            throw LeafError.getInstance().makeRuntimeError(
                expr.identifier().line(),
                `'${expr
                    .identifier()
                    .lexeme()}' can not be called on non class instance`
            )
        }

        const instance = caller as LeafInstance

        if (!instance.fields().has(expr.identifier().lexeme())) {
            throw LeafError.getInstance().makeRuntimeError(
                expr.identifier().line(),
                `field '${expr
                    .identifier()
                    .lexeme()}' does not exist on instance of class'${instance
                    .className()
                    .lexeme()}'`
            )
        }

        const right = this.evaluate(expr.right())
        instance.fields().set(expr.identifier().lexeme(), right)

        return right
    }

    visitClassExpr(expr: ClassExpr): IObj {
        return LeafClass.createInstance(
            expr.name(),
            expr.fields(),
            Array.from(expr.methods().entries()).reduce(
                (prev, [key, value]) => {
                    prev.set(
                        key,
                        LeafMethod.createInstance(
                            value.parameters(),
                            value.stmts(),
                            this._environment
                        )
                    )
                    return prev
                },
                new Map()
            )
        )
    }

    visitThisExpr(expr: ThisExpr): IObj {
        if (this._callableCtx === CallableCtx.NONE) {
            throw LeafError.getInstance().makeRuntimeError(
                expr.keyword().line(),
                "'this' can not be used outside of a function or method"
            )
        }
        return this.instance()
    }

    visitGroupingExpr(expr: GroupingExpr): IObj {
        return this.evaluate(expr.expr())
    }

    visitLiteralExpr(expr: LiteralExpr): IObj {
        switch (expr.token().type()) {
            case tokenType.NUMBER:
                return ObjNumber.createInstance(
                    parseFloat(expr.token().lexeme())
                )
            case tokenType.STRING:
                return ObjString.createInstance(expr.token().lexeme())
            case tokenType.TRUE:
                return ObjBool.createInstance(true)
            case tokenType.FALSE:
                return ObjBool.createInstance(false)
        }
        throw "undefined literal"
    }

    visitIdentifierExpr(expr: IdentifierExpr): IObj {
        return this._environment.get(expr.token())
    }

    visitNilExpr(expr: NilExpr): IObj {
        return ObjNil.createInstance()
    }

    // helper functions
    private continueWrapper(stmt: BlockStmt) {
        try {
            this.visitBlockStmt(stmt)
        } catch (err) {
            if ((err as Error).message !== continueStmt) {
                throw err
            }
        }
    }

    getCallLine(expr: IExpr) {
        if (expr instanceof IdentifierExpr) {
            return expr.token().line()
        }
        return (expr as CallExpr | GetExpr).line()
    }

    private callClass(leafClass: LeafClass, expr: CallExpr): IObj {
        if (!leafClass.methods().has("constructor")) {
            return LeafInstance.createInstance(
                leafClass.name(),
                utils.makeInstanceFieldsTable(leafClass.fields())
            )
        }

        const prevCtx = this._callableCtx
        this._callableCtx = CallableCtx.FUNC

        const method = leafClass.methods().get("constructor") as LeafMethod
        method.setType("constructor")
        method.setClassName(leafClass)

        const value = method.call(expr.args(), this, expr)

        this._callableCtx = prevCtx
        return value
    }
}
